### ISSUE TITLE
Add logging of requests and responses

### ISSUE THREAD

**Issue by vbauer** [hub4j/github-api#151]
Add possibility to log requests and responses.
It will help to understand a lot of cases when library does not work.

**Comment by KostyaSha** [comment]
Could you describe what problems that you had? There are some issues when logging wouldn't help you at all.

**Comment by vbauer** [comment]
I faced with the following problem:

github-api works fine if I send requests by portions (~200-1000 for one time), but it works very unstable when I make a lot of requests (~4000) for a short period of time (it hangs or throws strange NPE).

It is not critical issue for me, as I said, I just send request by portions, but it will be interesting to know the root of this problem. It looks like (sometimes) Github responds incorrectly and github-api does not process it correctly (or some problems with concurrency).

**Comment by KostyaSha** [comment]
This maybe related to #159 and #160 for NPE we will need stacktrace or any debug information. I can suggest run your tool under debugger and add Exception breakpoint.

**Comment by vbauer** [comment]
Ok, I'll try to reproduce this problems tomorrow.

**Comment by szpak** [comment]
When embedding in own tools (like maven/gradle plugin) it could be useful to help diagnose end user problems.

**Comment by rmetzger** [comment]
You _should_ be able to use OkHttp with Logging interceptors:

```java
Cache cache = new Cache(new File(cacheDir), cacheMB * 1024 * 1024);
        try {
            HttpLoggingInterceptor logging = new HttpLoggingInterceptor(new HttpLoggingInterceptor.Logger() {
                @Override
                public void log(String message) {
                    System.out.println("Intercepted " + message);
                }
            });
            logging.setLevel(HttpLoggingInterceptor.Level.HEADERS);

            OkHttpClient.Builder okHttpClient = new OkHttpClient.Builder();
            okHttpClient.addInterceptor(logging);
            okHttpClient.cache(cache);

            gitHub = GitHubBuilder.fromEnvironment().withPassword("", "")
                    .withConnector(new OkHttp3Connector(new OkUrlFactory(okHttpClient.build())))
                    .build();
```
This should, but does not work for me :)

**Comment by bitwiseman** [comment]
@rmetzger  
Logging interceptors: I think this is due to the way we use OkHttp.  Working toward upgrading.

**Comment by marcin-majewski-sonarsource** [comment]
@rmetzger Did you at the end manage to get it right? I have the same working example, and the logging interceptors are not called.

**Comment by rmetzger** [comment]
IIRC I didn't manage to fix this :( @bitwiseman's comment seem to confirm that this is currently not supported.

**Comment by bitwiseman** [comment]
@marcin-majewski-sonarsource 
There is basic logging in this library now: https://github.com/hub4j/github-api/blob/e66a72387e6fcd5762e0122f8da17d271d27a756/src/main/java/org/kohsuke/github/GitHubClient.java#L422-L425

But full logging requires moving to a non-`HttpUrlConnection` system (#1085).  We could update `ObsoleteUrlFactory` but that would require significant testing.

**Comment by bitwiseman** [comment]
Closed by #1290.  
Switch to `OkHttpGitHubConnector` to avoid `ObsoleteUrlFactory` and have full interceptor behavior.
